### PR TITLE
VP9 RTP parser: add robust bounds checks, fix pictureID and references, add unit tests

### PR DESCRIPTION
**Pull request by alexliyu7352**
注意:这个是codex自主修复, 不是我手动写的代码, 合并前请仔细验证

### Motivation

- Harden VP9 RTP payload parsing against malformed/truncated packets to avoid OOB reads and crashes.
- Fix incorrect assembly of extended `pictureID` and incorrect consumption of reference indices in flexible mode.
- Add focused unit tests to cover descriptor variants and edge cases so regressions are caught automatically.

### Description

- Change parser signature to `int parse(const unsigned char *data, int dataLength)` and replace raw pointer arithmetic with `readByte`/`skipBytes` helpers that track remaining bytes and enforce bounds.
- Fix `largePictureID` assembly by concatenating bytes correctly instead of using `ntohs` arithmetic, and ensure `additionalReferenceIdx` is set from the parsed byte; also limit flexible-mode reference chain to at most three references.
- Parse scalability-structure fields (`spatialLayers`, `hasResolution`, `hasGof`) with bounds checks and correctly read resolution entries and GOF reference indices; return `-1` on truncation.
- Add an extra safety check in `VP9RtpDecoder::decodeRtp` to verify the parsed header offset is within payload bounds, and add `tests/test_vp9_rtp.cpp` with many descriptor/edge-case cases exercising parsing logic.

### Testing

- Added unit test `tests/test_vp9_rtp.cpp` covering L/F/P combinations, picture ID variants, reference chains, resolution and GOF parsing, and truncation cases, and executed it as an automated test.
- The new test binary completed all test cases successfully (all expected pass/fail outcomes matched).

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a63e97a4c2c8320aa66d51a746c2e4d)

### CODE REVIEW

**Comment by xia-chu** [comment]
## PR #4790 审核报告

### 概述

此 PR 对 VP9 RTP 载荷解析器进行了安全加固，主要改动包括：用 `readByte`/`skipBytes` lambda 替代裸指针运算 + `VP9_CHECK_BOUNDS` 宏、修复 `largePictureID` 拼接逻辑错误、修复 flexible 模式下 Layer Indices 引用索引双重消费 bug、增加引用链上限校验（≤3 次，符合 RFC 7741）、解析失败时正确清理帧状态，以及新增 `tests/test_vp9_rtp.cpp` 单元测试。

---

### 正确修复的 Bug

**1. `largePictureID` 拼接逻辑**

旧代码 `ntohs((pictureID << 16) + byte)` 存在位移错误和不必要的字节序转换。新代码 `(pictureID << 8) | byte` 正确组装 15-bit Picture ID。

**2. Flexible 模式引用索引双重消费**

旧代码在 L=1, F=1, P=1 时，先在 `hasLayerIndices` 块内 do-while 循环消费引用索引，又在独立的 `flexibleMode && interPicturePrediction` 块中再次消费，导致解析位置错乱。新代码按 RFC 规范正确处理，L 块仅读取 T/U/S/D 字节（非 flexible 模式额外读取 `tl0PicIdx`），引用索引仅在 P,F 块中消费。

**3. 解析失败帧清理**

旧代码 parse 失败仅 `return false`，不清理已积累的帧数据，可能导致残留数据被拼入不完整帧。新代码正确设置 `_frame_drop = true` 并清空 `_frame->_buffer`。

**4. 引用链上限校验**

新增 `referenceCount > 3` 检查，符合 RFC 7741 "up to 3 reference indices" 的限制，防止畸形输入导致越界读取。

---

### 需要关注的问题

#### `remaining == 0` 检查过于严格，误拒合法的"仅描述符"RTP 包

`parse()` 末尾新增的 `if (remaining == 0) return -1;` 将"描述符恰好消耗完所有字节"视为解析失败。但根据 RFC 规范，VP9 RTP 包可以仅包含 payload descriptor 而不含 VP9 编码数据（例如多包帧组装中某个 RTP 包的 payload 恰好只够放下 descriptor）。

此时 `offset == payload_size`，`_frame->_buffer.append(payload + offset, 0)` 是无害的空操作。对比同项目的 VP8 解析器，其解析完成后没有类似的零 payload 拒绝逻辑。

同样，`decodeRtp` 中 `static_cast<size_t>(offset) >= payload_size` 使用 `>=` 而非 `>` 也拒绝了这一合法情况。

**建议修复**（已生成增量 patch `pr4790_fix_remaining_check.patch`）：

1. 删除 `parse()` 末尾的 `if (remaining == 0) return -1;`
2. 将 `decodeRtp` 中的 `>=` 改为 `>`：
   ```cpp
   if (offset < 0 || static_cast<size_t>(offset) > payload_size) {
   ```
3. 调整测试用例 `"descriptor only"` 和 `"layer descriptor only"` 的期望值，使其断言这些包应被成功接受：
   ```cpp
   runCase({ "descriptor only", { 0x00 }, 1 });
   runCase({ "layer descriptor only", { 0x30, 0x00 }, 3 });
   ```

> 注：实际场景中仅含描述符、无帧数据的 RTP 包确实少见，拒绝它作为防御性措施也有一定道理。此处建议取决于对 RFC 严格兼容性的要求。

---

### 测试代码评价

测试覆盖了 L/F/P 所有组合、Picture ID 变体（短/长/截断）、引用链（正常/截断/超限）、分辨率和 GOF 解析、以及解析失败后的帧恢复。`testParseFailureDropsCurrentFrame` 验证了帧状态清理修复，设计合理。测试质量较好。

---

### 总结

| 维度 | 评价 |
|------|------|
| Bug 修复正确性 | ✅ `largePictureID` 和 flexible 模式双重消费都是真实 bug |
| 安全性 | ✅ 边界检查全面加固 |
| 代码质量 | ✅ `readByte`/`skipBytes` 比宏+裸指针更清晰安全 |
| 测试覆盖 | ✅ 覆盖了关键路径和边界情况 |
| 待讨论 | ⚠️ `remaining == 0` 检查是否过于严格 |

总体来看，这是一个质量较高的 PR，核心的 bug 修复和安全加固都有价值。唯一建议调整的是 `remaining == 0` 的检查策略，详见上方增量 patch。

```patch
diff --git a/ext-codec/VP9Rtp.cpp b/ext-codec/VP9Rtp.cpp
index def02127..d121a238 100644
--- a/ext-codec/VP9Rtp.cpp
+++ b/ext-codec/VP9Rtp.cpp
@@ -208,7 +208,7 @@ int RTPPayloadVP9::parse(const unsigned char *data, int dataLength) {
       }
     }
   }
-  if (remaining == 0) return -1;
+
   return dataPtr - data;
 }
 
@@ -245,7 +245,7 @@ bool VP9RtpDecoder::decodeRtp(const RtpPacket::Ptr &rtp) {
 
     RTPPayloadVP9 info;
     int offset = info.parse(payload, payload_size);
-    if (offset < 0 || static_cast<size_t>(offset) >= payload_size) {
+    if (offset < 0 || static_cast<size_t>(offset) > payload_size) {
         WarnL << "VP9 RTP payload parse failed, seq:" << seq;
         _frame_drop = true;
         _frame->_buffer.clear();
diff --git a/tests/test_vp9_rtp.cpp b/tests/test_vp9_rtp.cpp
index 694d502b..df180b89 100644
--- a/tests/test_vp9_rtp.cpp
+++ b/tests/test_vp9_rtp.cpp
@@ -117,8 +117,8 @@ int main() {
     runCase({ "GOF exact boundary", { 0x02, 0x08, 0x02, 0x04, 0x11, 0x08, 0x22, 0x33, 0xAA }, 8 });
     runCase({ "GOF truncated references", { 0x02, 0x08, 0x01, 0x08, 0x22 }, -1 });
 
-    runCase({ "descriptor only", { 0x00 }, -1 });
-    runCase({ "layer descriptor only", { 0x30, 0x00 }, -1 });
+    runCase({ "descriptor only", { 0x00 }, 1 });
+    runCase({ "layer descriptor only", { 0x30, 0x00 }, 3 });
     testParseFailureDropsCurrentFrame();
     return 0;
 }
```

**Comment by xia-chu** [comment]
@alexliyu7352 上面千问3.7max的审核报告 你看看要不合并下patch

**Comment by alexliyu7352** [comment]
> @alexliyu7352 上面千问3.7max的审核报告 你看看要不合并下patch

这个需要仔细验证下, 当前只是看这个patch, 如果这样改可能会导致VP9FrameHelper::keyFrame() 的实现会对空 frame data 解引用. 具体还要更新下测试跑下. 现在看来AI对c/c++独立审核修复总是会有一些问题. 所以还需要交叉review一次看看. 

怎么感觉国产大模型比codex能打阿.:)

**Comment by alexliyu7352** [comment]
codex再次review的结论和我刚才说的判断大体类似, 认为patch可能会导致引入空帧读取.
以下是review的正文, 抱歉从codex 云端无法复制markdown过来,所以格式丢失了.

> 
> 对审核报告的逐项复核
> 报告对四项既有修复的判断是正确的
> 以下结论没有争议：
> 
> 扩展的图片身份证明需要按照 (high7 << 8) | low8 的方式进行组装；
> 
> L=1,F=1,P=1 的 reference chain 之前确实会被消费两次；
> 
> parse failure 后需要丢弃正在组装的帧；
> 
> 灵活模式的参考指数最多为三个。
> 
> 问题仅出在它最后建议的 descriptor-only patch。
> 
> 为什么不能直接采用报告中的 patch
> 1. patch 会把 descriptor-only 的 B=1,E=1 包送入空帧输出路径
> 测试的 runCase() 会对输入首字节执行：
> 
> packet[0] |= 0x0C;
> 也就是强制设置 B=1,E=1。
> 
> 对于：
> 
> runCase({ "descriptor only", { 0x00 }, 1 });
> 实际送入 decoder 的 payload 是：
> 
> 0x0C
> 如果按报告建议：
> 
> 删除 remaining == 0；
> 
> 将 decoder 的 >= payload_size 改成 > payload_size；
> 
> 那么该包会被接受，执行流程是：
> 
> B=1：清空 _frame->_buffer，设置 _frame_drop=false；
> 
> append 的长度为 0；
> 
> E=1 ：调用 outputFrame() ；
> 
> outputFrame() 最终可能调用 VP9FrameHelper::keyFrame()；
> 
> VP9FrameHelper::keyFrame() 会读取 frame data 的首字节。
> 
> 但此时 _frame->_buffer 是空的。VP9FrameHelper::keyFrame() 的实现会对空 frame data 解引用，因此报告所给 patch 不只是兼容性策略变化，还可能把当前被安全拒绝的输入引入空帧读取风险。
> 
> 换言之：
> 
> _frame->_buffer.append(payload + offset, 0) 本身确实是安全的，但不能据此推导整个 decoder 调用链是安全的。审核报告只检查了 append，没有继续检查 outputFrame() 和 VP9FrameHelper::keyFrame()。
> 
> 这也是该报告最重要的遗漏。
> 
> 2. layer descriptor only 的期望 offset 写错了
> 审核报告建议：
> 
> runCase({ "layer descriptor only", { 0x30, 0x00 }, 3 });
> 但这个 packet 只有两个字节。
> 
> 0x30 表示：
> 
> L=1
> 
> F=1
> 
> P=0
> 
> 解析内容只有：
> 
> 强制性描述符字节：1 字节；
> 
> 层索引字节：1 个字节。
> 
> 因为 F=1，不会读取 TL0PICIDX。所以正确 descriptor offset 是：
> 
> 2
> 不可能是 3。
> 
> 而且在报告建议的 decoder 校验下：
> 
> offset > payload_size
> 如果测试真的期待 parser 返回 3，而 payload_size 是 2，decoder 仍会拒绝它。也就是说，报告提供的测试修改与它自己的 decoder 修改互相矛盾。
> 
> 正确的“接受 descriptor-only”测试若真要写，应是：
> 
> runCase({ "layer descriptor only", { 0x30, 0x00 }, 2 });
> 但它随后仍会遇到前述空 frame 输出问题。
> 
> 3. “VP8 没有同样检查”不能证明 VP9 当前修改错误
> 对比 VP8 实现可以作为一致性参考，但不能单独证明 VP9 decoder 应接受 descriptor-only 输入。还必须检查：
> 
> VP8 decoder 是否会输出零长度 frame；
> 
> VP8 frame 的 keyFrame() 是否会读取空 buffer；
> 
> VP8 的 B/E 或分片状态与 VP9 是否完全相同；
> 
> packetizer 是否实际可能产生 descriptor-only 包；
> 
> 上层是否保证 RTP payload 至少包含一个编码数据字节。
> 
> 审核报告没有完成这些调用链验证。
> 
> 4. 原始需求明确要求拒绝 descriptor-only
> 本 PR 的明确验收条件之一就是：
> 
> 在解析完成时拒绝没有剩余 frame payload 的输入，并在 VP9RtpDecoder::decodeRtp 保留独立边界校验。
> 
> 所以当前代码的两项检查不是意外行为：
> 
> if (remaining == 0) {
>     return -1;
> }
> 以及：
> 
> if (offset < 0 || static_cast<size_t>(offset) >= payload_size) {
> 它们共同实现了明确需求。
> 
> 最终结论
> 审核报告内容	结论
> 四项核心 bug 修复正确	正确
> descriptor-only 是否可能值得讨论	可以讨论
> 当前 PR 的检查“误拒合法输入”	证据不足，且违背明确需求
> 直接删除 remaining == 0	不应采用
> 将 >= 改成 >	不应单独采用
> layer descriptor only 期望 offset 为 3	明确错误，正确 offset 是 2
> 建议 patch 整体是否安全	不安全；没有覆盖空帧输出及 keyFrame() 读取路径
> 因此，我确认看的是同一份报告；结论仍然是：不要应用报告附带的 patch。
> 
> 如果未来确实要支持 descriptor-only RTP 包，不能只删除两项校验，而需要同时定义并实现：

descriptor-only 且 E=0 时的状态行为；

descriptor-only 且 E=1 时是否允许结束帧；

B=1,E=1 的空帧是否应直接忽略；

outputFrame() 不得对空 buffer 调用需要读取首字节的 VP9 frame 方法；

对应测试 offset 应正确使用 1 和 2，而不是 1 和 3。

这会成为一个独立的行为兼容性改动，而不是当前安全修复 PR 中应顺带接受的小调整。

**Comment by xia-chu** [comment]
pr先合并吧
